Add Cloud Foundry routes as a source

dgrizzanti opened the pull request:
This is a proposal to add Routes from Cloud Foundry as an optional source.

Delivers a ubiquitous Developer Experience to use External DNS open source project for both Kubernetes as well as Cloud Foundry. This streamlines application publishing using single, automated  DNS management across multiple application platforms and multiple clouds. Currently, this functionality does not exist across multiple Cloud Foundry foundations (or AZs).

General idea here is that you would point external-dns to a Cloud Foundry foundation/endpoint and it would map all routes under all domains back to the DNS zone you would like to manage with external-dns. This allows for routes registered in Cloud Foundry to have their DNS managed in an external provider (Route 53, Google Cloud DNS, etc).

njuettner commented:
Thank you for your PR! 

You may want to fix formatting first:

pkg/apis/externaldns/types.go:186:33:warning: missing ',' before newline in composite literal (deadcode)
pkg/apis/externaldns/types.go:186:33:warning: missing ',' before newline in composite literal (golint)

Additionally to that you have to make sure `dep` has all your new importing packages otherwise the build will fail.

After fixing we'll have a look :slightly_smiling_face:.

dgrizzanti commented:
Thanks for taking a look @njuettner. I've fixed those 2 issues but getting stuck on something else with the build. It seems like it's not finding the new config options I've added even though they have been added to `source/store.go`. Looking at the version of the project that is pulled into the vendor directory during the build?

Let me know if I'm missing something. Thanks!

Raffo commented:
From the TravisCI output it looks like there is a missing method to implement in the test and other issues: 

```
# github.com/kubernetes-incubator/external-dns/source [github.com/kubernetes-incubator/external-dns/source.test]
source/route.go:41:31: rs.client.Config undefined (type *cfclient.Client has no field or method Config)
source/store_test.go:64:25: cannot use mockClientGenerator (type *MockClientGenerator) as type ClientGenerator in argument to ByNames:
	*MockClientGenerator does not implement ClientGenerator (missing CloudFoundryClient method)
source/store_test.go:73:25: cannot use mockClientGenerator (type *MockClientGenerator) as type ClientGenerator in argument to ByNames:
	*MockClientGenerator does not implement ClientGenerator (missing CloudFoundryClient method)
source/store_test.go:83:25: cannot use mockClientGenerator (type *MockClientGenerator) as type ClientGenerator in argument to ByNames:
	*MockClientGenerator does not implement ClientGenerator (missing CloudFoundryClient method)
source/store_test.go:92:19: cannot use mockClientGenerator (type *MockClientGenerator) as type ClientGenerator in argument to ByNames:
	*MockClientGenerator does not implement ClientGenerator (missing CloudFoundryClient method)
source/store_test.go:95:18: cannot use mockClientGenerator (type *MockClientGenerator) as type ClientGenerator in argument to ByNames:
	*MockClientGenerator does not implement ClientGenerator (missing CloudFoundryClient method)
source/store_test.go:98:18: cannot use mockClientGenerator (type *MockClientGenerator) as type ClientGenerator in argument to ByNames:
	*MockClientGenerator does not implement ClientGenerator (missing CloudFoundryClient method)
source/store_test.go:107:19: cannot use mockClientGenerator (type *MockClientGenerator) as type ClientGenerator in argument to ByNames:
	*MockClientGenerator does not implement ClientGenerator (missing CloudFoundryClient method)
FAILED   /home/travis/gopath/src/github.com/kubernetes-incubator/external-dns/vendor/github.com/kubernetes/repo-infra/verify/go-tools/verify-govet.sh	31s
The command "vendor/github.com/kubernetes/repo-infra/verify/verify-go-src.sh -v --rootdir $(pwd)" exited with 1.
```

Do you think you can work at fixing those issues?

dgrizzanti commented:
thanks @Raffo will work through these issues

dgrizzanti commented:
Issues are fixed but still missing a few tests - will work on adding those

Raffo commented:
@dgrizzanti is this ready for review?

dgrizzanti commented:
@Raffo yes - it should be all set. Only comment would be that I opened this PR more as a proposal for adding the feature. The code works, but I probably need to update the docs on how to use it/intention, etc. I wanted someone to at least give the 👍 on the approach prior putting together a full set of docs.

I'm fine with adding all of that after the fact in if you think I'm good but wanted to get through that checkpoint first.

Thanks!

njuettner commented:
@dgrizzanti we switched to `go modules`, could you please rebase from master?

dgrizzanti commented:
@njuettner rebased and updated with go modules

njuettner commented on source/store.go:
```diff
@@ -91,6 +98,27 @@ func (p *SingletonClientGenerator) IstioClient() (istiomodel.ConfigStore, error)
+		Username:   cfUsername,
+		Password:   cfPassword,
+	}
+	client, _ := cfclient.NewClient(c)
```
Always check errors

njuettner commented on source/store.go:
```diff
@@ -91,6 +98,27 @@ func (p *SingletonClientGenerator) IstioClient() (istiomodel.ConfigStore, error)
 
+// CloudFoundryClient generates a cf client if it was not created before
+func (p *SingletonClientGenerator) CloudFoundryClient(cfAPIEndpoint string, cfUsername string, cfPassword string) (*cfclient.Client, error) {
+	var err error
```
Why do you initialise the error before and not assigning it directly?

njuettner commented on source/store.go:
```diff
@@ -130,6 +158,12 @@ func BuildWithConfig(source string, p ClientGenerator, cfg *Config) (Source, err
 			return nil, err
 		}
 		return NewIstioGatewaySource(kubernetesClient, istioClient, cfg.IstioIngressGatewayServices, cfg.Namespace, cfg.AnnotationFilter, cfg.FQDNTemplate, cfg.CombineFQDNAndAnnotation, cfg.IgnoreHostnameAnnotation)
+	case "route":
```
Do not use route use something related to cloudfoundry

njuettner commented on source/route_test.go:
```diff
@@ -0,0 +1,42 @@
+}
+
+func (suite *RouteSuite) SetupTest() {
+
```
Just omit this line

njuettner replied:
Line 31

njuettner commented:
/lgtm
/approve
